c_src/srtp/srtp.c

```diff
@@ -0,0 +1,432 @@
+    srtp_crypto_policy_set_null_cipher_hmac_sha1_80(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "null_cipher_hmac_null") == 0) {
+    srtp_crypto_policy_set_null_cipher_hmac_null(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_cm_256_hmac_sha1_80") == 0) {
+    srtp_crypto_policy_set_aes_cm_256_hmac_sha1_80(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_cm_256_hmac_sha1_32") == 0) {
+    srtp_crypto_policy_set_aes_cm_256_hmac_sha1_32(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_cm_256_null_auth") == 0) {
+    srtp_crypto_policy_set_aes_cm_256_null_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_cm_192_hmac_sha1_80") == 0) {
+    srtp_crypto_policy_set_aes_cm_192_hmac_sha1_80(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_cm_192_hmac_sha1_32") == 0) {
+    srtp_crypto_policy_set_aes_cm_192_hmac_sha1_32(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_cm_192_null_auth") == 0) {
+    srtp_crypto_policy_set_aes_cm_192_null_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_gcm_128_8_auth") == 0) {
+    srtp_crypto_policy_set_aes_gcm_128_8_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_gcm_256_8_auth") == 0) {
+    srtp_crypto_policy_set_aes_gcm_256_8_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_gcm_128_8_only_auth") == 0) {
+    srtp_crypto_policy_set_aes_gcm_128_8_only_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_gcm_256_8_only_auth") == 0) {
+    srtp_crypto_policy_set_aes_gcm_256_8_only_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_gcm_128_16_auth") == 0) {
+    srtp_crypto_policy_set_aes_gcm_128_16_auth(policy);
+    return true;
+  }
+
+  if (strcmp(crypto_profile, "aes_gcm_256_16_auth") == 0) {
+    srtp_crypto_policy_set_aes_gcm_256_16_auth(policy);
+    return true;
+  }
+
+  return false;
+}
+
+UNIFEX_TERM create(UnifexEnv *env) {
+  UnifexState *state = unifex_alloc_state(env);
+  state->session = NULL;
+
+  srtp_err_status_t err = srtp_create(&state->session, NULL);
+  if (err) {
+    unifex_release_state(env, state);
+    return unifex_raise(env, srtp_strerror(err));
+  }
+
+  return create_result(env, state);
+}
+
+void handle_destroy_state(UnifexEnv *env, UnifexState *state) {
+  if (state->session) {
+    srtp_dealloc(state->session);
+    state->session = NULL;
+  }
+
+  unifex_release_state(env, state);
+}
+
+static void free_master_keys_array(srtp_policy_t *policy) {
+  if (!policy->keys && policy->num_master_keys > 0) {
+    for (size_t i = 0; i < policy->num_master_keys; i++) {
+      srtp_master_key_t *key = policy->keys[i];
+      enif_free(key);
+    }
+
+    enif_free(policy->keys);
+    policy->keys = NULL;
+    policy->num_master_keys = 0;
+  }
+}
+
+static UNIFEX_TERM
+create_master_keys_array(UnifexEnv *env, UnifexPayload **keys,
```

Comment:
can't the checks from this function be done on the elixir side?

Reply:
well they are based on data that is fetched from unexposed parts of the policy struct so they must be on C side (or the structure should be fully exposed in Elixir but this would require me to support structs, lists etc.)